From 90760ec0600661eda277b35d6fd85b933baf86e8 Mon Sep 17 00:00:00 2001
From: Monika Alekhya <malekh@codeaurora.org>
Date: Mon, 1 Apr 2013 18:20:20 +0530
Subject: [PATCH 164/176] Camera: Fix deadlock in pcb and snapshot threads

In non-zsl case of takepicture,we do streamoff for preview
stream which is waiting on preview callback thread to exit.
By that time the lock has already been acquired by takePicture.
So preivew callback will not exit until it acquires lock and
takePicture cannot continue until PCB call back is returned.

Change-Id: I11f7e841ba3382981c8199a0f116029a0feed0c4
CRs-Fixed: 468626
---
 QCamera/HAL/core/src/QCameraHWI_Preview.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/QCamera/HAL/core/src/QCameraHWI_Preview.cpp b/QCamera/HAL/core/src/QCameraHWI_Preview.cpp
index a63c062..a9cd2c9 100755
--- a/QCamera/HAL/core/src/QCameraHWI_Preview.cpp
+++ b/QCamera/HAL/core/src/QCameraHWI_Preview.cpp
@@ -726,6 +726,8 @@ status_t QCameraStream_preview::processPreviewFrameWithDisplay(mm_camera_super_b
       //Sending preview callback if corresponding Msgs are enabled
       if(mHalCamCtrl->mMsgEnabled & CAMERA_MSG_PREVIEW_FRAME) {
           ALOGE("Q%s: PCB callback enabled", __func__);
+      //if CAMERA_MSG_COMPRESSED_IMAGE is enabled then avoid preview callback to app to avoid deadlock scenarios.
+      if(!(mHalCamCtrl->mMsgEnabled & CAMERA_MSG_COMPRESSED_IMAGE)) {
           msgType |=  CAMERA_MSG_PREVIEW_FRAME;
           int previewBufSize;
           /* The preview buffer size sent back in the callback should be (width*height*bytes_per_pixel)
@@ -755,6 +757,7 @@ status_t QCameraStream_preview::processPreviewFrameWithDisplay(mm_camera_super_b
                 data = mHalCamCtrl->mPreviewMemory.camera_memory[preview_buf_idx];
                 ALOGE("Invalid preview format, buffer size in preview callback may be wrong.");
           }
+      }
       } else {
           data = NULL;
       }
-- 
1.8.3.1

